### DISCUSSION TITLE
[Raytracing in one weekend] 10.3 Total internal reflection

### DISCUSSION THREAD

**Discussion by irevoire** [RayTracing/raytracing.github.io#713] [Debugging]
Hello,

I’m following this book in rust to discover raytracing and I have a strange bug / artifact.
I was wondering if someone already had this kind of artifact and knew where it could come from.

Basically everything works great until the 10.3 part where I get a totally black sphere:
![test2](https://user-images.githubusercontent.com/7032172/91415956-04deeb80-e84f-11ea-910a-4f505d9dc891.png)

I tried to do only reflection or only refraction and both seems to works:
<img width="1440" alt="Capture d’écran 2020-08-26 à 23 10 02" src="https://user-images.githubusercontent.com/7032172/91416041-1cb66f80-e84f-11ea-9706-b470922f6816.png">

It breaks when I use this condition to choose which one to do: `if etai_over_etat * sin_theta > 1.`.

Also I tried to do the next part with the schlick approximation to see if it would fix the problem, it does not. Here is the result if it can help:

<img width="1440" alt="image" src="https://user-images.githubusercontent.com/7032172/91416273-7159ea80-e84f-11ea-91ea-9270d0ccbb2b.png">

My code is here is someone want to check something:
https://github.com/irevoire/raytracing_in_one_weekend

**Comment by L3tum** [comment]
I've had a similar problem with the "edges" of the sphere becoming black. The image seems to be correct *enough* without it so I just commented it out for now. I'd love to hear from you if you figure it out

**Comment by irevoire** [comment]
I really don’t see any difference between my code and the official code. Are you also using rust @L3tum ?
I’m wondering if it could be a difference in the way NaN are handled 🤔

**Comment by L3tum** [comment]
I'm actually using C#, but the NaN handling seems to be different in that `==` and such operators do not honor NaN, but the `.Equals` method does (which nobody uses). 
It may just be a bug in the official code as well. I made another issue regarding `FlipFace` in Book 2 and how I think it's missing. So let's hope someone picks this up and enlightens us^^

From my limited amount of googling it appears that there may be two bugs at hand. One is that the reflected ray will intersect the sphere immediately again and the other is that the refracted ray may *not* intersect through the sphere. [Here](https://stackoverflow.com/questions/16453450/raytracer-refraction-bug) is a stackoverflow question about it as well.
I'm not sure if it helps, yet. I'll probably start playing around with it in a few hours and see if any of it helps me.

**Comment by L3tum** [comment]
I've implemented the code only for the Total Internal Reflection from the answer of that stackoverflow question and it seems to work exactly the same. 

However while playing around I've noticed that using `r_in.origin` (ray.Origin) instead of `rec.p` (record.Position) for the scattered ray origin seems to fix it. It's obviously not really correct in terms of actual true to life imagery. 

```cs
if (etaiOverEtat * sinTheta > 1f)
{
    var reflected = Util.ReflectVec(unitDirection, record.Normal);
    scattered = new Ray(ray.Origin, reflected, ray.Time);
    return true;
}
```

[Here](https://imgur.com/a/BalRqSq) is an album showing the 3 implementations. 

I don't think it's applicable to your issue though since your entire sphere is black. I had the problem that some rectangles in the cornell box were black and found out that my normal method wasn't correct (I was using (0,0,0) as the ray direction when calculating the front face). Maybe that's your problem as well.
